automerge-wasm/README.md

```diff
@@ -1,696 +1,478 @@
+
+  let doc = create()
+
+  let config = doc.putObject("/", "config", { align: "left", archived: false, cycles: [10, 19, 21] })
+
+  doc.put(config, "align", "right")
+
+  // Anywhere Object Ids are being used a path can also be used.
+  // The following two statements are equivalent:
+
+  // get the id then use it
+
+  let id = doc.value("/", "config")[1]
+  if (id && id[0] === 'map') {
+    doc.put(id[1], "align", "right")
+  }
+
+  // use a path instead
+
+  doc.put("/config", "align", "right")
+
+  doc.free()
+```
+
+Using the id directly is always faster (as it prevents the path to id conversion internally) so it is preferred for performance critical code.
+
+### Maps
+
+Maps are key/value stores.  The root object is always a map.  The keys are always strings.  The values can be any scalar type or any object.
+
+```javascript
+    let doc = create()
+    let mymap = doc.putObject("_root", "mymap", { foo: "bar"})
+                              // make a new map with the foo key
+
+    doc.put(mymap, "bytes", new Uint8Array([1,2,3]))
+                              // assign a byte array to key `bytes` of the mymap object
+
+    let submap = doc.putObject(mymap, "sub", {})
+                              // make a new empty object and assign it to the key `sub` of mymap
+
+    doc.keys(mymap)           // returns ["bytes","foo","sub"]
+    doc.materialize("_root")  // returns { mymap: { bytes: new Uint8Array([1,2,3]), foo: "bar", sub: {}}}
+    doc.free()
+```
+
+### Lists
+
+Lists are index addressable sets of values.  These values can be any scalar or object type.  You can manipulate lists with `insert()`, `put()`, `insertObject()`, `pubObject()`, `push()`, `pushObject()`, `splice()`, and `del()`.
+
+```javascript
+    let doc = create()
+    let items = doc.putObject("_root", "items", [10,"box"])
+                                                  // init a new list with two elements
+    doc.push(items, true)                         // push `true` to the end of the list
+    doc.putObject(items, 0, { hello: "world" })  // overwrite the value 10 with an object with a key and value
+    doc.del(items, 1)                             // delete "box"
+    doc.splice(items, 2, 0, ["bag", "brick"])     // splice in "bag" and "brick" at position 2
+    doc.insert(items, 0, "bat")                   // insert "bat" to the beginning of the list
+    doc.insertObject(items, 1, [1,2])             // insert a list with 2 values at pos 1
+
+    doc.materialize(items)                        // returns [ "bat", [1,2], { hello : "world" }, true, "bag", "brick"]
+    doc.length(items)                             // returns 6
+    doc.free()
+```
+
+### Text
+
+Text is a specialized list type intended for modifying a text document.  The primary way to interact with a text document is via the slice operation.  Non text can be inserted into a text document and will be represented with the unicode object replacement character.
```

Comment:
Should we add something here around the way strings are split in javascript vs other platforms. Feels like this is a common footgun so worth documenting.